File: programmers/[programmers] 소수 찾기.py
def solution(numbers):
    answer = 0
    filt = [1]*(10000000)
    filt[0]=0
    filt[1]=0
    for n in range(2, len(filt)):
        if filt[n]==1:
            i = 2
            while n*i < 10000000:
                filt[n*i]=0
                i+=1

    def dfs(num, level):
        nonlocal memo, used
        if level>len(numbers): return
        if num not in memo and filt[num]==1:
            memo.append(num)
        for i in range(len(numbers)):
            if used[i]==1: continue
            used[i]=1
            dfs(num*10+int(numbers[i]), level+1)
            used[i]=0

    memo = []
    used=[0]*len(numbers)
    dfs(0, 0)
    answer = len(memo)
    return answer

File: programmers/[programmers] 소수 찾기.js
function solution(numbers) {
  let answer = 0;
  const cards = numbers.split("");
  const used = Array(numbers.length).fill(false);
  const dict = Array(10000000).fill(true);
  const memo = new Set();
  dict[0] = false;
  dict[1] = false;

  for (let i = 2; i < dict.length ** 0.5; i++) {
    let k = 2;
    while (i * k <= 9999999) {
      dict[i * k] = false;
      k += 1;
    }
  }

  function isPrimeNumber(num) {
    return dict[num];
  }

  function dfs(depth, cur) {
    const num = Number(cur);
    if (isPrimeNumber(num) && !memo.has(num)) {
      answer += 1;
      memo.add(num);
    }
    if (depth === numbers.length) return;
    for (let i = 0; i < numbers.length; i++) {
      if (used[i]) continue;
      used[i] = true;
      dfs(depth + 1, cur + numbers[i]);
      used[i] = false;
    }
  }
  dfs(0, "");
  return answer;
}
